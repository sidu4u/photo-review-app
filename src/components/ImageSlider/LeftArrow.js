import React from "react";
import PropTypes from 'prop-types';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import Button from "../../components/Button";
    

const LeftArrow = ({onClick})=>{
    return <Button>
        <ChevronLeftIcon id="left_arrow" onClick={onClick}/>
    </Button>
   }


LeftArrow.propType = {
    onClick : PropTypes.func
}   

export default LeftArrow;

