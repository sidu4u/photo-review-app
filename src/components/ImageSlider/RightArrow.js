import React from "react";
import PropType from 'prop-types';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import Button from "../../components/Button";


const RightArrow = ({onClick})=>{
    return <Button>
        <ChevronRightIcon id="right_arrow" onClick={onClick}/>
    </Button>
   }


RightArrow.propType = {
    onClick : PropType.func
}   

export default RightArrow;

