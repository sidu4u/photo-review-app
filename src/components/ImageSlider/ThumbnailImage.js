import React from "react";
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Image from '../Image';

const ThumbnailImage = styled(({className,src})=>{
    return <Image className={className} src={src}/>
 })`
 margin-right:10px;
 width:60px;
 height:30px;
 `;

 ThumbnailImage.propTypes={
    className:PropTypes.string,
    src:PropTypes.string
 }

 export default ThumbnailImage;