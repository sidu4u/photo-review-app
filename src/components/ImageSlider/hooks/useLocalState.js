import {useCallback,useState,useMemo} from "react";

const  useLocalState = (images)=>{
    const pages = useMemo(()=>Math.ceil(images.length/3),[images]);
    let [currentPageNumber,updateCurrentPageNumber] = useState(1);
    const nextPage = useCallback(()=>{
        updateCurrentPageNumber(currentPageNumber=>{
            if(currentPageNumber>=pages){
                return currentPageNumber;
            }
            return currentPageNumber+1
        });
    },[pages]);
    const previousPage = useCallback(()=>{
        updateCurrentPageNumber(currentPageNumber=>{
            if(currentPageNumber<=1){
                return currentPageNumber;
            }
            return currentPageNumber-1}); 
    },[]);

    return {currentPageNumber,nextPage,previousPage,pages};
}

export default useLocalState;