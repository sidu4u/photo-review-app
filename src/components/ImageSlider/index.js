import React from "react";
import styled from 'styled-components';
import ChevronRightIcon from '@material-ui/icons/ChevronRight';
import ChevronLeftIcon from '@material-ui/icons/ChevronLeft';
import ThumbnailImage from './ThumbnailImage';
import useLocalState from "./hooks/useLocalState";

const FlexDiv = styled.div`
display:flex;
`;
const StyledDiv = styled(FlexDiv)`
display:flex;
margin-top:10px;
`;
const OverflowDiv = styled.div`
width:300px;
overflow:hidden
`;

const ReviewedImageContainer = ({images})=>{
    const {currentPageNumber,previousPage,nextPage,pages} = useLocalState(images);
     return (
         <StyledDiv>
             {currentPageNumber>1&&<ChevronLeftIcon id="left_arrow" onClick={previousPage}/>}
            <OverflowDiv>
                <FlexDiv>
                {images.slice((currentPageNumber-1)*3,currentPageNumber*3).map(image=>(<ThumbnailImage id={image.id} key={image.id} src={image.urls.raw}/>))}
                </FlexDiv>
             </OverflowDiv>
             {currentPageNumber<=pages&& <ChevronRightIcon id="right_arrow" onClick={nextPage}/>}
        </StyledDiv>
     )
}

export default ReviewedImageContainer;