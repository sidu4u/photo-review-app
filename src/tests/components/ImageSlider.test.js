import React from "react";
import renderer from 'react-test-renderer';
import {shallow} from "enzyme";
import ImageSlider from '../../components/ImageSlider';
import useLocalState from "../../components/ImageSlider/hooks/useLocalState";

const images = [{id:1},{id:2},{id:3},{id:4},{id:5},{id:6},{id:7}].map(({id})=>({id,urls:{raw:`${id}`}}))

it('renders correctly',()=>{
    const component = renderer
    .create(<ImageSlider images={images}/>)
    .toJSON();
    expect(component).toMatchSnapshot();
});

function HookWrapper({images}) {
    const hook = useLocalState(images,3);
    return <div hook={hook} />;
}

it('renders with first page as default',()=>{
  let wrapper = shallow(<HookWrapper images={images}/>);
  let { hook } = wrapper.find('div').props();
  let {currentPageNumber}= hook;
  
  expect(currentPageNumber).toBe(1);

});

it('page number should not go below 1',()=>{
    let wrapper = shallow(<HookWrapper images={images}/>);
    let {hook:{previousPage}} = wrapper.find('div').props();
    previousPage();
    let {hook:{currentPageNumber}} = wrapper.find('div').props();
    expect(currentPageNumber).toBe(1);
  });


it('page number should increase on next click',()=>{
    let wrapper = shallow(<HookWrapper images={images}/>);
    let {hook:{nextPage}} = wrapper.find('div').props();
    nextPage();
    let {hook:{currentPageNumber}} = wrapper.find('div').props();
    expect(currentPageNumber).toBe(2);
  }); 
  
  it('page number should decrease on previous click',()=>{
    let wrapper = shallow(<HookWrapper images={images}/>);
    let {hook:{nextPage,previousPage}} = wrapper.find('div').props();
    nextPage();
    previousPage();
    let {hook:{currentPageNumber}} = wrapper.find('div').props();
    expect(currentPageNumber).toBe(1);
  }); 
   
  it('page number should not be more than number of pages',()=>{
    let wrapper = shallow(<HookWrapper images={images}/>);
    let {hook:{nextPage,pages}} = wrapper.find('div').props();
    
    for(let i=1;i<=pages+2;i++){  // increaseing more than number of pages
        nextPage();
    }
     
    let {hook:{currentPageNumber}} = wrapper.find('div').props();
    expect(currentPageNumber).toBe(pages);
  }); 
   

