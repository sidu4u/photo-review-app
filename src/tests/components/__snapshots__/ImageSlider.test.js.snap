// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-gtsrHT sc-dlnjwi eHuwNW uguGA"
>
  <div
    className="sc-hKFxyN erwvSU"
  >
    <div
      className="sc-gtsrHT eHuwNW"
    >
      <img
        alt=""
        className="sc-bdnxRM jBrNAE"
        src="1"
      />
      <img
        alt=""
        className="sc-bdnxRM jBrNAE"
        src="2"
      />
      <img
        alt=""
        className="sc-bdnxRM jBrNAE"
        src="3"
      />
    </div>
  </div>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root"
    focusable="false"
    id="right_arrow"
    onClick={[Function]}
    viewBox="0 0 24 24"
  >
    <path
      d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
    />
  </svg>
</div>
`;
